concordium-consensus/src/Concordium/Scheduler.hs

```diff
@@ -1101,79 +1102,61 @@ handleDeployCredential ::
-                        -- we checked regIdEx above already).
-                        accExistsAlready <- isJust <$> lift (getAccount aaddr)
-                        let check = AH.verifyCredential cryptoParams ipInfo arsInfos cdiBytes (Left messageExpiry)
-                        unless (not accExistsAlready && check) $ throwError $ Just AccountCredentialInvalid
-                        -- Add the account to the state, but only if the credential was valid and the account does not exist
-                        _ <- lift (createAccount cryptoParams aaddr cdv)
-                        mkSummary (TxSuccess [AccountCreated aaddr, CredentialDeployed{ecdRegId=regId,ecdAccount=aaddr}])
-  case res of
-    Left err -> return (TxInvalid <$> err)
-    Right ts -> return (Just ts)
-
+handleDeployCredential accCreation@AccountCreation{messageExpiry=messageExpiry, credential=cdi} cdiHash = do
+    res <- runExceptT $ do
+      liftedCm <- lift getChainMetadata
+      let ts = slotTime liftedCm
+      -- check that the transaction is not expired
+      when (transactionExpired messageExpiry ts) $ throwError (Just ExpiredTransaction)
+      remainingEnergy <- lift getRemainingEnergy
+      let cost = Cost.deployCredential (ID.credentialType cdi) (ID.credNumKeys . ID.credPubKeys $ cdi)
+      when (remainingEnergy < cost) $ throwError Nothing
+      let mkSummary tsResult = do
+            tsIndex <- lift bumpTransactionIndex
+            return $ TxValid $ TransactionSummary{
+              tsSender = Nothing,
+              tsHash = cdiHash,
+              tsCost = 0,
+              tsEnergyCost = cost,
+              tsType = TSTCredentialDeploymentTransaction (ID.credentialType cdi),
+              ..
+              }
+
+      let regId = ID.credId accCreation
+      let aaddr = ID.addressFromRegId regId
+      -- We always need to make sure that the account was not created in between
+      -- the transaction was received and the actual execution.
+      accExistsAlready <- isJust <$> lift (getAccount aaddr)
+      when accExistsAlready $ throwError $ Just $ DuplicateAccountRegistrationID regId
+      liftedCryptoParams <- lift TV.getCryptographicParameters      
+      cachedTVResult <- lift (lookupTransactionVerificationResult cdiHash)
+      case cachedTVResult of
+        Just TV.Success -> do
+          return ()
+        Just _ -> do 
+          -- we verify the transaction again as it might have become
+          -- valid since it was last verified.
+          tVerResult <- lift (TV.verifyCredentialDeployment ts accCreation)
+          when (tVerResult /= TV.Success) $ throwError $ mapErr tVerResult
+        Nothing -> do
+          -- If the transaction has not been verified before we verify it now
+          tVerResult <- lift (TV.verifyCredentialDeployment ts accCreation)
+          when (tVerResult /= TV.Success) $ throwError $ mapErr tVerResult
```

Comment:
No, as subsequent transactions with the same hash - i.e., duplicates will not be forwarded to the consensus.

Reply:
But we could have to execute the same transaction multiple times since it could be in multiple blocks. That should be relatively rare, but not unheard of.